ext/curl/interface.c

```diff
@@ -63,6 +63,16 @@
 #include "curl_private.h"
 #include "curl_arginfo.h"
 
+#if (HAVE_SOCKETS || defined(COMPILE_DL_SOCKETS))
```

Comment:
I don't think this will work if ext/sockets is compiled separately as a shared extension...

Reply:
I haven't looked much into the implementation yet, but I'm wondering if it might make sense to make this based on (fd-castable) streams rather than sockets...

Reply:
It doesn't work as a shared extension. That's why the windows build is failing. I'm used to just build it as static and then everything works fine ofc.
I was trying to find a way around it, but I shouldn't use socket_ce as it will not exist when the sockets extension isn't loaded.

The only way I found for now was to strcmp the ce->name of the retval to make sure I got a Socket on the opensocket callback and not auto-create a socket object on the other two callbacks when the zval isn't set.
That would imply you can only use sockopt and closesocket when you created a socket using socket_create in the opensocket callback.I think that's a fair compromise as most use-cases will have an opensocket callback anyway when using the closesocket.

Using streams won't make it easier I think. Curl expects to get a socket that isn't connected yet by default. When you do the callback you don't have the port curl is going to connect to so you'd have to rely on users to have the STREAM_CLIENT_CONNECT flag not set.

Reply:
> Using streams won't make it easier I think. Curl expects to get a socket that isn't connected yet by default. When you do the callback you don't have the port curl is going to connect to so you'd have to rely on users to have the STREAM_CLIENT_CONNECT flag not set.

I'm primarily suggesting to use streams because they are always available, so we don't have to deal with the interaction with ext/sockets, which probably won't have any really great solution. Streams and sockets can be converted into each other (see https://www.php.net/socket_export_stream), so this is not a restriction in power.

It looks like the connection can be handled either by calling stream_socket_client without STREAM_CLIENT_CONNECT, or by returning CURL_SOCKOPT_ALREADY_CONNECTED from the sockopt callback. TBH I don't fully understand why these two callbacks are separate.